Rework the plugin system

mbasaglia opened the pull request:
Implements most of the things mentioned here: https://github.com/synfig/synfig/issues/1309#issuecomment-606829467

ice0 commented:
It's ok, what travis build is errored, i will fix it ASAP.

ice0 commented:
Ok, Travis build fixed. After rebase all should be fine.

mbasaglia commented:
I've added support for importers defined in plugins.

To test this you can use my plugin: http://mattia.basaglia.gitlab.io/tgs/downloads.html
It supports a couple different formats, probably the easiest to test is SVG.

morevnaproject commented:
@mbasaglia Thank you for an awesome addition! Please check my comments here - https://github.com/synfig/synfig/issues/1309#issuecomment-607334235 ^__^

morevnaproject commented:
I have tested this PR, here are my thoughts:

1. We now have two "Import" menu commands:
![screenshot_001](https://user-images.githubusercontent.com/332868/78243869-c139bd00-750e-11ea-89a0-dca6fbdd05e0.png)

I was unable to test properly "Import file..." command, because importer gave me an error when I made an attempt to import Lottie json file. But that's different issue.

I guess that "Import file..." opens imported file in a new tab (not adds to currently opened document). So, I think the meaning of it is more like "File" - "Open" command.

I think that we should merge "Import file..." and "Open" commands for better UX.

2. Export command works perfectly, I have tested with html and json export. ^__^

mbasaglia commented:
yeah it does work more like open, I'll move the code to that dialog. 

to test import the easiest way is to create a simple drawing in synfig, export with "[TGS] SVG" and import the same file.

morevnaproject commented:
Thank you! I want to say that I am absolutely excited about this feature and which possibilities it  opens for further development.

For example, a very common case when newbie user tries to import some image using "File" - "Open" command. Synfig give them error, because images intended to be added via "File" - "import" command. But newbie users are not aware about that and they often reach out to support asking what's wrong.

With new plugins system we can create a very simple plugin, which allows to open images via "File" - "Open" command. The plugin will do simple thing - it will create a new document with imported image. ^__^

mbasaglia commented:
I've pushed some changed that do that and also fixed some issues.

mbasaglia commented:
Yea

mbasaglia commented:
CI fail seems unrelated to my changes

```
$ ./1-setup-osx-brew.sh

Checking adwaita-icon-theme...

==> Downloading https://homebrew.bintray.com/bottles/gnome-icon-theme-3.22.0.yos

Error: Failed to download resource "gnome-icon-theme"

Download failed: https://homebrew.bintray.com/bottles/gnome-icon-theme-3.22.0.yosemite.bottle.tar.gz

[...]

==> Downloading https://download.gnome.org/sources/adwaita-icon-theme/3.22/adwai

Error: Failed to download resource "gnome-icon-theme"

Download failed: https://download.gnome.org/sources/adwaita-icon-theme/3.22/adwaita-icon-theme-3.22.0.tar.xz

The command "./1-setup-osx-brew.sh" failed and exited with 1 during .
```

AnishGG commented:
I have restarted the job. Sometimes the download fails because of network errors at the server.

morevnaproject commented:
> I've pushed some changes that do that and also fixed some issues.

Thank you! I will be able to test it on Tuesday. ^__^

mbasaglia commented:
By the way, as a side effect of this PR, #1252 should be much easier to fix as all that is required is to recreate / update the menu items when language is changed.
(Previously plugin names were loaded at startup and not using the correct language).

morevnaproject commented:
Hi! I did one more round of testing and found the following:

The Lottie plugin (which is only capable for export currently) for some reason is displayed in "File" - "Open" dialog. The plugin doesn't has any importers declared, so I guess it shouldn't be there. ^__^ 

(I do not have any other export/import plugins installed, besides Lottie, which is shipped with this version of Synfig).

![screenshot_001](https://user-images.githubusercontent.com/332868/78693087-f5edb000-7924-11ea-89a1-e6c43a42067e.png)

morevnaproject commented:
Also, I suggest to make "Open history" button inactive when any filetype handled by plugin is selected.
![screenshot_003](https://user-images.githubusercontent.com/332868/78693756-e6bb3200-7925-11ea-8e93-b7aeb7419649.png)

mbasaglia commented:
> Also, I suggest to make "Open history" button inactive when any filetype handled by plugin is selected.
> ![screenshot_003](https://user-images.githubusercontent.com/332868/78693756-e6bb3200-7925-11ea-8e93-b7aeb7419649.png)

I'm not too familiar with Gtk(mm) and I wasn't able to find a signal for when the filter changes (doesn't help that the documentation isn't great) do you know how that's done?

https://developer.gnome.org/gtkmm/stable/classGtk_1_1FileChooserDialog.html

rodolforg commented:
As far as I can tell, you can't access the filter combobox (and its signals) in a normal way, only what filter is currently active.

So I think "Open history" should behave differently according to the selected filter – if it's possible.

Regarding the Gtkmm docs, I use it, but often I read the Gtk docs themselves, as they're more ordered and cleaner than `mm` counterpart.

mbasaglia commented:
> As far as I can tell, you can't access the filter combobox (and its signals) in a normal way, only what filter is currently active.
> 
> So I think "Open history" should behave differently according to the selected filter – if it's possible.

Currently for plugins "open history" just acts like open. 

So I guess it isn't possible to disable the open history button, even if it's possible to write the code that enables/disables based on the current filter, it can't be run when the filter changes if we can't hook to the combo signal.

rodolforg commented:
Couldn't the "Open history" action callback check what is the selected filter instead of  (sadly) disabling the button?

mbasaglia commented:
I'm not sure I follow, what should happen if someone clicks it when a plugin importer is selected?

mbasaglia commented:
This is the relevant code btw, clicking on "Open history" just sets an output variable when clicked from the dialog

https://github.com/synfig/synfig/blob/742010d34f33049693e3496dd76d7d7460cd8053/synfig-studio/src/gui/app.cpp#L2970

And this is where it's used, looks like it only does something for ".sfg" files? So clicking on .sif, .sifz doesn't do anything special either (from what I gather)

https://github.com/synfig/synfig/blob/742010d34f33049693e3496dd76d7d7460cd8053/synfig-studio/src/gui/app.cpp#L4196-L4238

morevnaproject commented:
> And this is where it's used, looks like it only does something for ".sfg" files?

Yes, the "Open history" works for ".sfg" format only.

I am not insist on reworking behavior of "Open history" button right now, in this exact PR. Just brought that up that maybe this could be fixed in the future if not now. ^__^

I am testing the PR again now. Thanks to everyone! ^__^

morevnaproject commented:
I found the crash is happening now with current code.
To reproduce it I have installed TGS plugin (as it is providing importers, we need them to trigger the crash).

Also, you will need 2 Lottie JSON (or any other importable) files to reproduce the crash.
I have used those ones - 
[lottie-2.zip](https://github.com/synfig/synfig/files/4464135/lottie-2.zip)

1. In Synfig choose "File" - "Open".
2. Navigate to directory where you have saved your Lottie JSON files.
3. From filetype dropdown choose "[TGS} Lottie JSON" (otherwise you will not see json files in open dialog)
![screenshot_006](https://user-images.githubusercontent.com/332868/79036982-a0045b00-7bf7-11ea-9a6e-3e40bcbcd2c2.png)
4. Choose first file "ball-1.json" and click open.
5. File will be opened.
6. After that choose "File" - "Open" again from menu
7. Notice that filetype dropdown has empty value now. I guess this is the reason of the bug.
![screenshot_007](https://user-images.githubusercontent.com/332868/79037021-e9ed4100-7bf7-11ea-8586-4841e8c3ca2e.png)
8. Select "ball-1.json" and click Open button.
9. Synfig crashes.

morevnaproject commented:
Video demonstration for issue -  https://youtu.be/7Po9pR-fg5M

morevnaproject commented:
Issue 2:

1. Launch Synfig, choose "File" - "Open".
2. From filetype dropdown choose "Any files"
![screenshot_010](https://user-images.githubusercontent.com/332868/79037247-5d438280-7bf9-11ea-916f-bd8d2ea4c2a3.png)
3. Choose file "ball-1.json" and click Open.
4. Error appears (because Synfig tries to read file as SIF).
![screenshot_008](https://user-images.githubusercontent.com/332868/79037256-7a785100-7bf9-11ea-9b1d-2c6b192b85d8.png)

morevnaproject commented:
Finally, suggestion:
It would be nice to have "All supported formats" option in file type drowpdown box and have it selected by default in "Open FIle" dialog. Example how it looks in Krita:
![screenshot_011](https://user-images.githubusercontent.com/332868/79037287-ce833580-7bf9-11ea-8019-d3038da6be16.png)

Then user will immediately see all files that can be opened by Synfig and that's a big advantage. ^__^

mbasaglia commented:
One thing to consider regarding selecting the plugin with "All supported formats" is what to do if two different plugins handle the same extension.

The easy solution would be to select the first one, but this choice might not be intuitive.

Maybe show a second dialog with a combo box if there are multiple matches?

mbasaglia commented:
I've set the dialog to use "Any files" by default, this also fixes the blank entry.
I've also added support to load plugins from "Any files" (just selects the first one for now).

One thing about the "open history" button, maybe it can be removed and we can add another entry in the filter combo? It might be more intuitive since it's an operation specific to a single file type.

morevnaproject commented:
> Maybe show a second dialog with a combo box if there are multiple matches?

I like this idea!

mbasaglia commented:
I've implemented the dialog to disambiguate plugins.

With this feature I've also added support for open recent (so imported files are listed there, and if there's ambiguity the dialog is shown)

morevnaproject commented:
Awesome!
![screenshot_003](https://user-images.githubusercontent.com/332868/79044815-9697e480-7c31-11ea-9097-4de6b658ae7b.png)

morevnaproject commented:
> I've set the dialog to use "Any files" by default, this also fixes the blank entry.

Can we have "All supported types" instead "Any files" by default?
I mean that with "Any files" we have many files visible, which cannot be opened. This can be confusing. ^__^"

Also, I found one more issue:
1. Launch Synfig and choose "File" - "Open".
2. Select any SVG file (i.e. tiger. Click Open.
3. File is loaded.
4. Make some edits.
5. Choose "File" - "Save". File marked as saved. Note that the name of saved file is "tiger.svg".
![screenshot_005](https://user-images.githubusercontent.com/332868/79045000-e9be6700-7c32-11ea-99cf-13a25a06c212.png)
6. Close file.
7. Choose "File" - "Open". And open same file again.
8. File opened without changes. That means no changes were saved at step 5.

I think the solution would be to ask to save into Synfig format at step 5 - display "Save As..." dialog (just like GIMP does).

mbasaglia commented:
I've added the extra file filter and made synfig handle files opened with a plugin in a similar way to how it handles new files. 

So for these files it will still show tiger.svg after opening, but when saving it recognizes it doesn't have a proper file name set and will prompt the file dialog (after which the name on the tab will be updated with the name of the saved file)

morevnaproject commented:
> So for these files it will still show tiger.svg after opening, but when saving it recognizes it doesn't have a proper file name set

Awesome, this is exactly what's needed. Testing now.
Thank you!

morevnaproject commented:
Testing complete. I think it is ready to merge now.

I haven't looked at the code, though. @ice0 or @rodolforg can you please take a look before merge? ^__^

rodolforg commented on synfig-studio/src/gui/app.cpp:
```diff
@@ -1210,15 +1210,11 @@ DEFINE_ACTION("keyframe-properties", _("Properties"));
-	list<PluginManager::plugin> plugin_list = studio::App::plugin_manager.get_list();
-	for(list<PluginManager::plugin>::const_iterator p=plugin_list.begin();p!=plugin_list.end();++p) {
-
+	for ( const auto& p : studio::App::plugin_manager.plugins() ) {
```
Just a cosmetic change for readability: keep "plugin" name instead of "p"

rodolforg commented on synfig-studio/src/gui/canvasview.cpp:
```diff
@@ -1517,14 +1517,13 @@ CanvasView::init_menus()
-	std::list<PluginManager::plugin> plugin_list = App::plugin_manager.get_list();
     auto instance = get_instance().get();
-	for(std::list<PluginManager::plugin>::const_iterator p = plugin_list.begin(); p != plugin_list.end(); ++p)
+	for ( const auto& p : App::plugin_manager.plugins() )
```
again, pls better naming :) "plugin" even for this short code slice

rodolforg commented on synfig-studio/src/gui/instance.cpp:
```diff
@@ -359,10 +260,13 @@ studio::Instance::run_plugin(std::string plugin_path, bool modify_canvas, std::v
 	String filename_processed;
+	String filename_prefix;
+	if ( !is_absolute_path(filename_original) )
+		filename_prefix = temporary_filesystem->get_temporary_directory() + ETL_DIRECTORY_SEPARATOR;
```
Someone can check if this temporary directory works fine on Windows?
If it points to a regular (=not admin) user has write access on the returned folder?
It could not be inside the folder Synfig is installed, but a system-defined one.
@mbasaglia @ice0

mbasaglia replied:
on my system it doesn't even use /tmp, it uses ~/.sinfig/tmp

rodolforg replied:
My worry is on MS Windows :(

I had a bad experience on using it installed on a Windows PC  by IT crew. I couldn't save the file :O

mbasaglia replied:
I think the filesystem gets constructed using `App::get_temporary_directory()` which is defined like that:
```c++
App::get_temporary_directory()
{
	return synfigapp::Main::get_user_app_directory() + ETL_DIRECTORY_SEPARATOR + "tmp";
}
```

mbasaglia replied:
by the way, `temporary_filesystem->get_temporary_directory()` was already used in this function

rodolforg replied:
I don't think it's a reason to not merge this commit.
It just sounds a bell on my rings :)

morevnaproject replied:
No worries, this dir is in `$HOME/Synfig` on Windows. It is proven to work as it is used for running plugins and autosave functionality. ^__^

rodolforg replied:
Well, I did had problems with 1.3.11
https://github.com/synfig/synfig/issues/1052

Actually, that's why I started to contributed to Synfig... tried to cleanup etl... in order to understand code. But I'm not a regular Windows user and building is too slow for Windows. So, I started to do other things... lol
Someday I gave up and opened the issue

morevnaproject commented:
@mbasaglia Can you please rebase on top of current master? ^__^

mbasaglia commented:
> @mbasaglia Can you please rebase on top of current master? ^__^

After rebase invoking plugins causes a segmentation fault on `stream_in->rdbuf();` in `run_plugin`

mbasaglia commented:
It happens on master as well

rodolforg commented:
Did you check what was the merge conflict on rebase?
This could help to find the commit that is probably guilty

mbasaglia commented:
Conflict was just some argument being changed from `string` to `std::string`, as I mentioned I get the same issue on a `master` checkout

rodolforg commented:
Oh
Not plugin related? :/
(And I meant the guilty commit in master branch)

mbasaglia commented:
it affects run_plugin, but it seems regardless of my changes

mbasaglia commented:
hmm, it seems to only affect some files

rodolforg commented:
Can you check if until commit 6e6eee54232375bfd30dca43b4a4a745d7d21d43 it still worked?

If so, try to `git bisect` from this commit on ;)

mbasaglia commented:
Found a fix

mbasaglia commented:
um, I re-based again. Any chance of this getting merged?

ice0 commented:
Sorry. Was busy last days.

Merged. Thank you!

mbasaglia commented:
Cool, I'll be documenting the new stuff in the XML file on the wiki
